hera_qm/xrfi.py

```diff
@@ -1,51 +1,116 @@
     Returns:
         array: filtered array. Same shape as input array.
-    """
+    '''
     d_sm = np.empty_like(d)
     for i in xrange(d.shape[0]):
         for j in xrange(d.shape[1]):
-            i0, j0 = max(0, i - K), max(0, j - K)
-            i1, j1 = min(d.shape[0], i + K), min(d.shape[1], j + K)
+            i0, j0 = max(0, i - Kt), max(0, j - Kf)
+            i1, j1 = min(d.shape[0], i + Kt), min(d.shape[1], j + Kf)
             d_sm[i, j] = medmin(d[i0:i1, j0:j1])
     return d_sm
 
+def detrend_deriv(d, dt=True, df=True):
+    '''XXX This only works ok on sparse RFI.'''
+    if df:
+        d_df = np.empty_like(d)
+        d_df[:, 1:-1] = (d[:, 1:-1] - .5 * (d[:, :-2] + d[:, 2:])) / np.sqrt(1.5)
+        d_df[:, 0] = (d[:, 0] - d[:, 1]) / np.sqrt(2)
+        d_df[:, -1] = (d[:, -1] - d[:, -2]) / np.sqrt(2)
+    else:
+        d_df = d
+    if dt:
+        d_dt = np.empty_like(d_df)
+        d_dt[1:-1] = (d_df[1:-1] - .5 * (d_df[:-2] + d_df[2:])) / np.sqrt(1.5)
+        d_dt[0] = (d_df[0] - d_df[1]) / np.sqrt(2)
+        d_dt[-1] = (d_df[-1] - d_df[-2]) / np.sqrt(2)
+    else:
+        d_dt = d
+    d2 = np.abs(d_dt)**2
+    # model sig as separable function of 2 axes                                                                                                                 
+    sig_f = np.median(d2, axis=0)
+    sig_f.shape = (1, -1)
+    sig_t = np.median(d2, axis=1)
+    sig_t.shape = (-1, 1)
+    sig = np.sqrt(sig_f * sig_t / np.median(sig_t))
+    return d_dt / sig
+
+
+def detrend_medminfilt(d, Kt=8, Kf=8):
+    """Detrend array using medminfilt statistic. See medminfilt.
+    Args:
+        d (array): data array of the shape (time, frequency) to detrend   
+        Kt (int): size in time to apply medminfilter over
+        Kf (int): size in frequency to apply medminfilter over
+    Returns:        
+         bool array: boolean array of flags    
+    """
+    d_sm = medminfilt(np.abs(d), 2 * Kt + 1, 2 * Kf + 1)
+    d_rs = d - d_sm
+    d_sq = np.abs(d_rs)**2
+    # puts minmed on same scale as average
+    if Kt == Kf:
+        sig = np.sqrt(medminfilt(d_sq, 2 * Kt + 1)) * (Kt / .64)
+    else:
+        sig = np.sqrt(medminfilt(d_sq, 2 * Kt + 1, 2 * Kf + 1)) * (np.sqrt(Kt**2 + Kf**2) / .64)
+    f = d_rs / sig
+    return f
+
+def detrend_medfilt(d, Kt=8, Kf=8):
+    """Detrend array using a median filter.    
+    Args:        
+        d (array): data array to detrend.        
+        K (int, optional): box size to apply medminfilt over    
+    Returns:        
+        bool array: boolean array of flags    
+    """
+    d = np.concatenate([d[Kt - 1::-1], d, d[:-Kt - 1:-1]], axis=0)
+    d = np.concatenate([d[:, Kf - 1::-1], d, d[:, :-Kf - 1:-1]], axis=1)
+    d_sm = medfilt(d, kernel_size=(2 * Kt + 1, 2 * Kf + 1))
+    d_rs = d - d_sm
+    d_sq = np.abs(d_rs)**2
+    # puts median on same scale as average    
+    # .456 scaling?    
```

Comment:
Yeah, I don't know where this number comes from. @jsdillon is this similar to the scaling you use in a modified z-score?